src/main/java/org/jboss/ejb/client/EJBClientContext.java

```diff
@@ -1148,51 +1155,76 @@ private void waitForClusterTopology(final String clusterName) {
 
     private void attemptReconnections() {
         final CountDownLatch reconnectTasksCompletionNotifierLatch;
-        synchronized(this) {
+        final List<ReconnectHandler> reconnectHandlersToAttempt;
+
+        // check for circumstances where we don't want to attempt reconnection
+        synchronized (this) {
```

Comment:
It looks like we're combining synchronization, Locks, and atomic/CAS.  Is this intended, or is there a simpler reduction of this code possible?